graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java

```diff
@@ -75,7 +77,12 @@ public DefaultGraphQLInvocation(
             executionInputBuilder.dataLoaderRegistry(dataLoaderRegistry.get());
         }
         ExecutionInput executionInput = executionInputBuilder.build();
-        return Flowable.fromPublisher(graphQLExecutionInputCustomizer.customize(executionInput, httpRequest))
+
+        Flowable<ExecutionInput> executionInputFlowable = graphQLExecutionInputCustomizer != null
```

Comment:
Why is this added?  The `GraphQLExecutionInputCustomizer`  should never be `null` as we use a `DefaultGraphQLExecutionInputCustomizer` ? Did you encounter any issues with current approach?

Reply:
Sometimes it can be null, I have encountered this issue. I thought it was a bug, but when I opened this file I saw the constructor allow null values here:

```java
    public DefaultGraphQLInvocation(
            GraphQL graphQL,
            @Nullable GraphQLExecutionInputCustomizer graphQLExecutionInputCustomizer,
            @Nullable Provider<DataLoaderRegistry> dataLoaderRegistry) {
```

That's why I have added that checking.

I can provide an example when `graphQLExecutionInputCustomizer` can be `null`.

Reply:
Interesting. As the `DefaultGraphQLExecutionInputCustomizer` has `@Singleton` annotation it should always have that one, unless replaced by the user.
You don't have to work out a complete example, but it's worth investigating how it happened in your case.

Reply:
Here is an example: https://github.com/donbeave/micronaut-graphql-customizer-null. I think if switch to use `Provider`, it will solve such error probably.

Reply:
Yeah, using `Provider` can fix that NPE issue: https://github.com/micronaut-projects/micronaut-graphql/pull/99/commits/c0a0c4c66842e4c468896040c78ae19ebc2b1d91

Reply:
Did you find out why it was `null` without the provider, the code seemed to be ok so I wonder what was going on.
I understand `Provider` fixes it but imo it seems unnecessary as the Micronaut-GraphQL should always have a  `GraphQLExecutionInputCustomizer`; either a custom one or the `DefaultGraphQLExecutionInputCustomizer`.

Reply:
`null` will be if there are multiple `Singleton` classes with the same interface `GraphQLExecutionInputCustomizer`, but none of them marked with `Primary` annotation, which is what my example project shows. Of course, this is a developer issue, I think, and the developer must add `Primary` annotation always in such cases, but some can forget (which was I did), and with NPE error thought it was not an issue in their code, but probably in the plugin code, that's why Provider looks better for me in such scenarios.

Reply:
With using `Provider` the error will be much more clearer:

```
Caused by: io.micronaut.context.exceptions.NonUniqueBeanException: Multiple possible bean candidates found: [io.micronaut.configuration.graphql.DefaultGraphQLExecutionInputCustomizer, example.graphql.RequestResponseCustomizer]
```

Reply:
Interesting, you should indeed use @Replaces on the ` RequestResponseCustomizer` class.
Maybe this is something that can be improved in Micronaut in general to have clearer error message?
@graemerocher @ilopmar is it recommended to inject a `Provider<GraphQLExecutionInputCustomizer>`  even if we know the Micronaut-GraphQL extensions always supplies a `DefaultGraphQLExecutionInputCustomizer`? I don't think it's good to do this only for getting a better error message, so I would say we should keep it without `Provider`.

Reply:
Agree, I also think it's better to have the same error with/or without `Provider`.

Reply:
`GraphQLExecutionInputCustomizer` shouldn't be `Nullable` if there is always an instance created by the library. The `package-info.java` file is annotated with `@RequiresGraphQL`, which requires a `GraphQL` bean and the property `graphql.enabled` not being `false`. Only in case that those conditions are false, the bean won't be created. But none of the beans in the package will be created either.

The library doesn't include a factory to create the `GraphQL` bean, and according to the docs, its seems the users need to do it. So again, I don't see how that bean can be `null` when the "requirements" for creating the beans are true.

Reply:
@ilopmar I think the issue @donbeave was experiencing when there are 2 `GraphQLExecutionInputCustomizer` beans. In that case the error was not clear. 
I think the right approach for now is to remove using the `Provider`  in this PR @donbeave .

Reply:
I see... When a user replaces a Micronaut bean it should add `@Primary` to use it as you both have been discussing. 
I've seen the "multiple beans candidates" some times and the error has been clear without using a `Provider`.

Btw, another option is to annotate the `DefaultGraphQLExecutionInputCustomizer` with `@Requires(missingBeans = {GraphQLExecutionInputCustomizer.class})` so it will only be created if the user doesn't provide its own bean.

Reply:
The `null` value will be only if there 2 or more `GraphQLExecutionInputCustomizer` beans already and none of them marked with `Primary` annotation. If use provider than it will not be nullable, but instead when try to get bean it will throw a `NonUniqueBeanException` exception.

Ok, @marceloverdijk, I'll remove `Provider` from here.

Reply:
And probably a good addition what @ilopmar mentioned, that should also cause no duplicate `GraphQLExecutionInputCustomizer` beans.

Reply:
> The `null` value will be only if there 2 or more `GraphQLExecutionInputCustomizer` beans already and none of them marked with `Primary` annotation. If use provider than it will not be nullable, but instead when try to get bean it will throw a `NonUniqueBeanException` exception.

If you remove the `@Nullable` annotation from the constructor it will fail with the "multiple bean candidates" messages when there are two beans because Micronaut won't know which one to inject. And with this you won't need to add any additional `@Requires` or mark the bean as `@Primary`. Only the user will need to do it, but they will know because the error message will be clear.

Reply:
@marceloverdijk I'll add it too.

Reply:
Done here: https://github.com/micronaut-projects/micronaut-graphql/pull/99/commits/d74bc324f09a13b162ea9e43e920a914df17f6e0

Reply:
@ilopmar got it, thank you for explanation how it works